File: Java/src/com/yourtion/Pattern05/HigherOrderFunctions.java
package com.yourtion.Pattern05;

import java.util.*;

/**
 * Created by Yourtion on 11/10/2016.
 */
public class HigherOrderFunctions {

    public static Integer sumSequence(List<Integer> sequence) {
        Integer sum = 0;
        for (Integer num : sequence) {
            sum += num;
        }
        return sum;
    }


    public static List<String> prependHello(List<String> names) {
        List<String> prepended = new ArrayList<String>();
        for (String name : names) {
            prepended.add("Hello " + name);
        }
        return prepended;
    }


    public static Set<Character> vowelsInWord(String word) {

        Set<Character> vowelsInWord = new HashSet<Character>();

        for (Character character : word.toLowerCase().toCharArray()) {
            if (isVowel(character)) {
                vowelsInWord.add(character);
            }
        }

        return vowelsInWord;
    }

    private static Boolean isVowel(Character c){
        Set<Character> vowels = new HashSet<Character>();
        vowels.add('a');
        vowels.add('e');
        vowels.add('i');
        vowels.add('o');
        vowels.add('u');
        return vowels.contains(c);
    }

    public void run() {
        System.out.println("vowelsInWord: " + vowelsInWord("Yourtion"));
        System.out.println();

        List<String> nameList =  Arrays.asList("Yourtion", "Sophia");
        System.out.println("person: " + nameList);
        System.out.println("prependHello: " + prependHello(nameList));
        System.out.println();

        List<Integer> sequence = Arrays.asList(1, 2, 3, 4, 5);
        System.out.println("sumSequence: " + sumSequence(sequence));
    }

}

File: Scala/src/com/yourtion/Pattern05/HigherOrderFunctions.scala
package com.yourtion.Pattern05

/**
  * Created by Yourtion on 11/10/2016.
  */
object HigherOrderFunctions {

  def sumSequence(sequence : Seq[Int]) =
    if(sequence.isEmpty) 0 else sequence.reduce((acc, curr) => acc + curr)

  def prependHello(names : Seq[String]) =
    names.map((name) => "Hello " + name)

  val isVowel = Set('a', 'e', 'i', 'o', 'u')
  def vowelsInWord(word: String) = word.filter(isVowel).toSet

  def run(): Unit = {
    println("vowelsInWord: " + vowelsInWord("Yourtion"))
    println()

    val nameList = Vector("Yourtion", "Sophia")
    println("person: " + nameList)
    println("prependHello: " + prependHello(nameList))
    println()

    val sequence = Vector(1, 2, 3, 4, 5)
    println("sumSequence: "+ sumSequence(sequence))
  }

}
